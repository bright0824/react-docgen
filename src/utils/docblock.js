/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

/**
 * Helper functions to work with docblock comments.
 */

var DOCLET_PATTERN = /^@(\w+)(?:$|\s((?:[^](?!^@\w))*))/gmi;

function parseDocblock(str) {
  var lines = str.split('\n');
  for (var i = 0, l = lines.length; i < l; i++) {
    lines[i] = lines[i].replace(/^\s*\*\s?/, '');
  }
  return lines.join('\n').trim();
}

/**
 * Given a path, this function returns the closest preceding docblock if it
 * exists.
 */
export function getDocblock(path: NodePath): ?string {
  if (path.node.comments) {
    var comments = path.node.comments.filter(function(comment) {
      return comment.leading &&
        comment.type === 'CommentBlock' &&
        comment.value.indexOf('*\n') === 0;
    });
    if (comments.length > 0) {
      return parseDocblock(comments[comments.length - 1].value);
    }
  }
  return null;
}

/**
 * Given a string, this functions returns an object with doclet names as keys
 * and their "content" as values.
 */
export function getDoclets(str: string): Object {
  var doclets = Object.create(null);
  var match = DOCLET_PATTERN.exec(str);

  for (; match; match = DOCLET_PATTERN.exec(str)) {
    doclets[match[1]] = match[2] || true;
  }

  return doclets;
}
