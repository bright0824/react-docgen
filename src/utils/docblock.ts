/**
 * Helper functions to work with docblock comments.
 */

import { CommentKind } from 'ast-types/gen/kinds';
import type { NodePath } from 'ast-types/lib/node-path';

const DOCLET_PATTERN = /^@(\w+)(?:$|\s((?:[^](?!^@\w))*))/gim;

function parseDocblock(str: string): string {
  // Does not use \s in the regex as this would match also \n and conflicts
  // with windows line endings.
  return str.replace(/^[ \t]*\*[ \t]?/gm, '').trim();
}

const DOCBLOCK_HEADER = /^\*\s/;

/**
 * Given a path, this function returns the closest preceding docblock if it
 * exists.
 */
export function getDocblock(path: NodePath, trailing = false): string | null {
  let comments: CommentKind[] = [];
  if (trailing && path.node.trailingComments) {
    comments = path.node.trailingComments.filter(
      comment =>
        comment.type === 'CommentBlock' && DOCBLOCK_HEADER.test(comment.value),
    );
  } else if (path.node.leadingComments) {
    comments = path.node.leadingComments.filter(
      comment =>
        comment.type === 'CommentBlock' && DOCBLOCK_HEADER.test(comment.value),
    );
  } else if (path.node.comments) {
    comments = path.node.comments.filter(
      comment =>
        comment.leading &&
        comment.type === 'CommentBlock' &&
        DOCBLOCK_HEADER.test(comment.value),
    );
  }

  if (comments.length > 0) {
    return parseDocblock(comments[comments.length - 1].value);
  }
  return null;
}

/**
 * Given a string, this functions returns an object with doclet names as keys
 * and their "content" as values.
 */
export function getDoclets(str: string): Record<string, unknown> {
  const doclets = Object.create(null);
  let match = DOCLET_PATTERN.exec(str);

  for (; match; match = DOCLET_PATTERN.exec(str)) {
    doclets[match[1]] = match[2] || true;
  }

  return doclets;
}
