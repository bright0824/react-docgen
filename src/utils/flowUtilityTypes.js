/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */
import recast from 'recast';

const {
  types: { namedTypes: types },
} = recast;

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(path: NodePath): boolean {
  if (types.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return Boolean(idPath) && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  return path.get('typeParameters', 'params', 0);
}
