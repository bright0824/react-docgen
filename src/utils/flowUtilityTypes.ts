import { namedTypes as t } from 'ast-types';
import type { NodePath } from 'ast-types/lib/node-path';

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(path: NodePath): boolean {
  if (t.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return !!idPath && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  while (isSupportedUtilityType(path)) {
    path = path.get('typeParameters', 'params', 0);
  }

  return path;
}
