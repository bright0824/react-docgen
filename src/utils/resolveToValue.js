/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import recast from 'recast';

var {
  types: {
    NodePath,
    builders,
    namedTypes: types,
  },
} = recast;

function buildMemberExpressionFromPattern(path: NodePath): ?NodePath {
  var node = path.node;
  if (types.Property.check(node)) {
    var objPath = buildMemberExpressionFromPattern(path.parent);
    if (objPath) {
      return new NodePath(
        builders.memberExpression(
          objPath.node,
          node.key,
          types.Literal.check(node.key)
        ),
        objPath
      );
    }
  } else if (types.ObjectPattern.check(node)) {
    return buildMemberExpressionFromPattern(path.parent);
  } else if (types.VariableDeclarator.check(node)) {
    return path.get('init');
  }
  return null;
}

function findScopePath(paths: Array<NodePath>, path: NodePath): ?NodePath {
  if (paths.length < 1) {
    return;
  }
  let resultPath = paths[0];
  const parentPath = resultPath.parent;

  if (types.ImportDefaultSpecifier.check(parentPath.node) ||
    types.ImportSpecifier.check(parentPath.node) ||
    types.ImportNamespaceSpecifier.check(parentPath.node) ||
    types.VariableDeclarator.check(parentPath.node) ||
    types.TypeAlias.check(parentPath.node)
  ) {
    resultPath = parentPath;
  } else if (types.Property.check(parentPath.node)) {
    // must be inside a pattern
    const memberExpressionPath = buildMemberExpressionFromPattern(
      parentPath
    );
    if (memberExpressionPath) {
      return memberExpressionPath;
    }
  }

  if (resultPath.node !== path.node) {
    return resolveToValue(resultPath);
  }
}

/**
 * If the path is an identifier, it is resolved in the scope chain.
 * If it is an assignment expression, it resolves to the right hand side.
 *
 * Else the path itself is returned.
 */
export default function resolveToValue(path: NodePath): NodePath {
  var node = path.node;
  if (types.VariableDeclarator.check(node)) {
     if (node.init) {
       return resolveToValue(path.get('init'));
     }
  } else if (
    types.ImportDefaultSpecifier.check(node) ||
    types.ImportNamespaceSpecifier.check(node) ||
    types.ImportSpecifier.check(node)
  ) {
    return path.parentPath;
  } else if (types.AssignmentExpression.check(node)) {
    if (node.operator === '=') {
      return resolveToValue(node.get('right'));
    }
  } else if (types.Identifier.check(node)) {
    if ((types.ClassDeclaration.check(path.parentPath.node) ||
        types.ClassExpression.check(path.parentPath.node) ||
        types.Function.check(path.parentPath.node)) &&
        path.parentPath.get('id') === path) {
      return path.parentPath;
    }

    let scope = path.scope.lookup(node.name);
    let resolvedPath: ?NodePath;
    if (scope) {
      const bindings = scope.getBindings()[node.name];
      resolvedPath = findScopePath(bindings, path);
    } else {
      scope = path.scope.lookupType(node.name);
      if (scope) {
        const types = scope.getTypes()[node.name];
        resolvedPath = findScopePath(types, path);
      }
    }
    return resolvedPath || path;
  }

  return path;
}
