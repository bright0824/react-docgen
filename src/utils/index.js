/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import * as docblock from './docblock';

export {
  docblock,
};

export {default as getClassMemberValuePath} from './getClassMemberValuePath';
export {default as getMemberExpressionRoot} from './getMemberExpressionRoot';
export {default as getMembers} from './getMembers';
export {default as getMemberValuePath} from './getMemberValuePath';
export {default as getNameOrValue} from './getNameOrValue';
export {default as getPropertName} from './getPropertyName';
export {default as getPropertyValuePath} from './getPropertyValuePath';
export {default as getPropType} from './getPropType';
export {default as isExportsOrModuleAssignment} from './isExportsOrModuleAssignment';
export {default as isReactComponentClass} from './isReactComponentClass.js';
export {default as isReactCreateClassCall} from './isReactCreateClassCall';
export {default as isReactModuleName} from './isReactModuleName';
export {default as match} from './match';
export {default as normalizeClassDefiniton} from './normalizeClassDefinition';
export {default as printValue} from './printValue';
export {default as resolveExportDeclaration} from './resolveExportDeclaration';
export {default as resolveToModule} from './resolveToModule';
export {default as resolveToValue} from './resolveToValue';
