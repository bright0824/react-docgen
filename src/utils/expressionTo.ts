/*eslint no-loop-func: 0, no-use-before-define: 0*/

import { namedTypes as t } from 'ast-types';
import resolveToValue from './resolveToValue';
import type { Importer } from '../parse';
import type { NodePath } from 'ast-types/lib/node-path';

/**
 * Splits a MemberExpression or CallExpression into parts.
 * E.g. foo.bar.baz becomes ['foo', 'bar', 'baz']
 */
function toArray(path: NodePath, importer: Importer): string[] {
  const parts = [path];
  let result: string[] = [];

  while (parts.length > 0) {
    path = parts.shift() as NodePath;
    const node = path.node;
    if (t.CallExpression.check(node)) {
      parts.push(path.get('callee'));
      continue;
    } else if (t.MemberExpression.check(node)) {
      parts.push(path.get('object'));
      if (node.computed) {
        const resolvedPath = resolveToValue(path.get('property'), importer);
        if (resolvedPath !== undefined) {
          result = result.concat(toArray(resolvedPath, importer));
        } else {
          result.push('<computed>');
        }
      } else {
        // @ts-ignore
        result.push(node.property.name);
      }
      continue;
    } else if (t.Identifier.check(node)) {
      result.push(node.name);
      continue;
    } else if (t.Literal.check(node)) {
      // @ts-ignore
      result.push(node.raw);
      continue;
    } else if (t.FunctionExpression.check(node)) {
      result.push('<function>');
      continue;
    } else if (t.ThisExpression.check(node)) {
      result.push('this');
      continue;
    } else if (t.ObjectExpression.check(node)) {
      const properties = path.get('properties').map(function (property) {
        if (t.SpreadElement.check(property.node)) {
          return `...${toString(property.get('argument'), importer)}`;
        } else {
          return (
            toString(property.get('key'), importer) +
            ': ' +
            toString(property.get('value'), importer)
          );
        }
      });
      result.push('{' + properties.join(', ') + '}');
      continue;
    } else if (t.ArrayExpression.check(node)) {
      result.push(
        '[' +
          path
            .get('elements')
            .map(function (el) {
              return toString(el, importer);
            })
            .join(', ') +
          ']',
      );
      continue;
    }
  }

  return result.reverse();
}

/**
 * Creates a string representation of a member expression.
 */
function toString(path: NodePath, importer: Importer): string {
  return toArray(path, importer).join('.');
}

export { toString as String, toArray as Array };
