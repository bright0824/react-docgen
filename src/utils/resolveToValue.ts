import { Scope } from '@babel/traverse';
import type { NodePath } from '@babel/traverse';
import type {
  Identifier,
  ImportDeclaration,
  MemberExpression,
} from '@babel/types';
import getMemberExpressionRoot from './getMemberExpressionRoot';
import getPropertyValuePath from './getPropertyValuePath';
import { Array as toArray } from './expressionTo';
import { traverseShallow } from './traverse';
import getMemberValuePath, {
  isSupportedDefinitionType,
} from './getMemberValuePath';
import initialize from './ts-types';

function findScopePath(
  bindingIdentifiers: Array<NodePath<Identifier>>,
): NodePath | null {
  if (bindingIdentifiers && bindingIdentifiers.length >= 1) {
    const resolvedParentPath = bindingIdentifiers[0].parentPath;

    if (
      resolvedParentPath.isImportDefaultSpecifier() ||
      resolvedParentPath.isImportSpecifier()
    ) {
      // TODO TESTME
      let exportName: string | undefined;

      if (resolvedParentPath.isImportDefaultSpecifier()) {
        exportName = 'default';
      } else {
        const imported = resolvedParentPath.get('imported');

        if (imported.isStringLiteral()) {
          exportName = imported.node.value;
        } else if (imported.isIdentifier()) {
          exportName = imported.node.name;
        }
      }

      if (!exportName) {
        throw new Error('Could not detect export name');
      }

      const importedPath = resolvedParentPath.hub.import(
        resolvedParentPath.parentPath as NodePath<ImportDeclaration>,
        exportName,
      );

      if (importedPath) {
        return resolveToValue(importedPath);
      }
    }

    return resolveToValue(resolvedParentPath);
  }

  return null;
}

/**
 * Tries to find the last value assigned to `name` in the scope created by
 * `scope`. We are not descending into any statements (blocks).
 */
function findLastAssignedValue(
  path: NodePath,
  idPath: NodePath<Identifier>,
): NodePath | null {
  const results: NodePath[] = [];
  const name = idPath.node.name;

  traverseShallow(path, {
    AssignmentExpression(assignmentPath) {
      const left = assignmentPath.get('left');

      // Skip anything that is not an assignment to a variable with the
      // passed name.
      // Ensure the LHS isn't the reference we're trying to resolve.
      if (
        !left.isIdentifier() ||
        left.node === idPath.node ||
        left.node.name !== name ||
        assignmentPath.node.operator !== '='
      ) {
        return;
      }
      // Ensure the RHS doesn't contain the reference we're trying to resolve.
      const candidatePath = assignmentPath.get('right');

      if (
        candidatePath.node === idPath.node ||
        idPath.findParent(parent => parent.node === candidatePath.node)
      ) {
        return;
      }

      results.push(candidatePath);

      return assignmentPath.skip();
    },
  });

  const resultPath = results.pop();

  if (resultPath == null) {
    return null;
  }

  return resolveToValue(resultPath);
}

/**
 * If the path is an identifier, it is resolved in the scope chain.
 * If it is an assignment expression, it resolves to the right hand side.
 * If it is a member expression it is resolved to it's initialization value.
 *
 * Else the path itself is returned.
 */
export default function resolveToValue(path: NodePath): NodePath {
  if (path.isIdentifier()) {
    if (
      (path.parentPath.isClass() || path.parentPath.isFunction()) &&
      path.parentPath.get('id') === path
    ) {
      return path.parentPath;
    }

    const binding = path.scope.getBinding(path.node.name);
    let resolvedPath: NodePath | null = null;

    if (binding) {
      // The variable may be assigned a different value after initialization.
      // We are first trying to find all assignments to the variable in the
      // block where it is defined (i.e. we are not traversing into statements)
      resolvedPath = findLastAssignedValue(binding.scope.path, path);
      if (!resolvedPath) {
        // @ts-ignore TODO fix in DT
        const bindingMap = binding.path.getOuterBindingIdentifierPaths(
          true,
        ) as Record<string, Array<NodePath<Identifier>>>;

        resolvedPath = findScopePath(bindingMap[path.node.name]);
      }
    } else {
      // Initialize our monkey-patching of @babel/traverse ðŸ™ˆ
      initialize(Scope);
      const typeBinding = path.scope.getTypeBinding(path.node.name);

      if (typeBinding) {
        resolvedPath = findScopePath([typeBinding.identifierPath]);
      }
    }

    return resolvedPath || path;
  } else if (path.isVariableDeclarator()) {
    const init = path.get('init');

    if (init.hasNode()) {
      return resolveToValue(init);
    }
  } else if (path.isMemberExpression()) {
    const root = getMemberExpressionRoot(path);
    const resolved = resolveToValue(root);

    if (resolved.isObjectExpression()) {
      let propertyPath: NodePath | null = resolved;

      for (const propertyName of toArray(path).slice(1)) {
        if (propertyPath && propertyPath.isObjectExpression()) {
          propertyPath = getPropertyValuePath(propertyPath, propertyName);
        }
        if (!propertyPath) {
          return path;
        }
        propertyPath = resolveToValue(propertyPath);
      }

      return propertyPath;
    } else if (isSupportedDefinitionType(resolved)) {
      const property = path.get('property');

      if (property.isIdentifier() || property.isStringLiteral()) {
        const memberPath = getMemberValuePath(
          resolved,
          property.isIdentifier() ? property.node.name : property.node.value, // TODO TESTME
        );

        if (memberPath) {
          return resolveToValue(memberPath);
        }
      }
    } else if (resolved.isImportDeclaration() && resolved.node.specifiers) {
      // Handle references to namespace imports, e.g. import * as foo from 'bar'.
      // Try to find a specifier that matches the root of the member expression, and
      // find the export that matches the property name.
      for (const specifier of resolved.get('specifiers')) {
        if (
          specifier.isImportNamespaceSpecifier() &&
          specifier.node.local &&
          specifier.node.local.name === (root.node as Identifier).name // TODO TESTME could be not an identifier
        ) {
          const resolvedPath = path.hub.import(
            resolved,
            ((root.parentPath.node as MemberExpression).property as Identifier)
              .name, // TODO TESTME Idk what that is
          );

          if (resolvedPath) {
            return resolveToValue(resolvedPath);
          }
        }
      }
    }
  } else if (
    path.isImportDefaultSpecifier() ||
    path.isImportNamespaceSpecifier() ||
    path.isImportSpecifier()
  ) {
    // go up to the import declaration
    return path.parentPath;
  } else if (
    path.isTypeCastExpression() ||
    path.isTSAsExpression() ||
    path.isTSTypeAssertion()
  ) {
    return resolveToValue(path.get('expression') as NodePath);
  }

  return path;
}
