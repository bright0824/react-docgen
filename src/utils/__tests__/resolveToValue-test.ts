import type { NodePath } from '@babel/traverse';
import {
  ExportNamedDeclaration,
  identifier,
  memberExpression,
  numericLiteral,
  objectProperty,
  restElement,
} from '@babel/types';
import type {
  AssignmentExpression,
  CallExpression,
  Identifier,
} from '@babel/types';
import { makeMockImporter, parse, parseTypescript } from '../../../tests/utils';
import resolveToValue from '../resolveToValue';

describe('resolveToValue', () => {
  const mockImporter = makeMockImporter({
    Foo: stmtLast =>
      stmtLast<ExportNamedDeclaration>(`
        const baz = 3;
        export { baz };
      `)
        .get('specifiers')[0]
        .get('local') as NodePath,
  });

  it('resolves simple variable declarations', () => {
    const path = parse.expressionLast(['var foo  = 42;', 'foo;'].join('\n'));

    expect(resolveToValue(path)).toEqualASTNode(numericLiteral(42));
  });

  it('resolves object destructuring', () => {
    const path = parse.expressionLast(
      ['var {foo: {bar: baz}} = bar;', 'baz;'].join('\n'),
    );

    const expected = objectProperty(
      identifier('bar'),
      identifier('baz'),
      undefined,
      undefined,
      undefined,
    );

    expected.decorators = undefined;
    // @ts-ignore BABEL types bug
    expected.method = false;
    // Resolves to identifier in destructuring
    expect(resolveToValue(path)).toEqualASTNode(expected);
  });

  it('handles RestElements properly', () => {
    const path = parse.expressionLast(
      ['var {foo: {bar}, ...baz} = bar;', 'baz;'].join('\n'),
    );

    expect(resolveToValue(path)).toEqualASTNode(restElement(identifier('baz')));
  });

  it('returns the original path if it cannot be resolved', () => {
    const path = parse.expressionLast(
      ['function foo() {}', 'foo()'].join('\n'),
    );

    expect(resolveToValue(path)).toEqualASTNode(path);
  });

  it('resolves variable declarators to their init value', () => {
    const path = parse.statement('var foo = 42;').get('declarations')[0];

    expect(resolveToValue(path)).toEqualASTNode(numericLiteral(42));
  });

  it('resolves to class declarations', () => {
    const path = parse.expressionLast(`
      class Foo {}
      Foo;
    `);

    expect(resolveToValue(path).node.type).toBe('ClassDeclaration');
  });

  it('resolves to class function declaration', () => {
    const path = parse.expressionLast(`
      function foo() {}
      foo;
    `);

    expect(resolveToValue(path).node.type).toBe('FunctionDeclaration');
  });

  describe('flow', () => {
    it('resolves type cast expressions', () => {
      const path = parse.expressionLast(`
      function foo() {}
      (foo: any);
    `);

      expect(resolveToValue(path).node.type).toBe('FunctionDeclaration');
    });
  });

  describe('typescript', () => {
    it('resolves type as expressions', () => {
      const path = parseTypescript.expressionLast(`
      function foo() {}
      (foo as any);
    `);

      expect(resolveToValue(path).node.type).toBe('FunctionDeclaration');
    });

    it('resolves type assertions', () => {
      const path = parseTypescript.expressionLast(`
      function foo() {}
      (<any> foo);
    `);

      expect(resolveToValue(path).node.type).toBe('FunctionDeclaration');
    });

    it('resolves type alias', () => {
      const path = parseTypescript.statement(
        `let action: Action;
         type Action = {};`,
      );

      expect(
        resolveToValue(
          path.get(
            'declarations.0.id.typeAnnotation.typeAnnotation.typeName',
          ) as NodePath,
        ).node.type,
      ).toBe('TSTypeAliasDeclaration');
    });
  });

  describe('assignments', () => {
    it('resolves to assigned values', () => {
      const path = parse.expressionLast(
        ['var foo;', 'foo = 42;', 'foo;'].join('\n'),
      );

      expect(resolveToValue(path)).toEqualASTNode(numericLiteral(42));
    });

    it('resolves to other assigned value if ref is in an assignment lhs', () => {
      const path = parse.expressionLast<AssignmentExpression>(
        ['var foo;', 'foo = 42;', 'foo = wrap(foo);'].join('\n'),
      );

      expect(resolveToValue(path.get('left'))).toEqualASTNode(
        numericLiteral(42),
      );
    });

    it('resolves to other assigned value if ref is in an assignment rhs', () => {
      const path = parse.expressionLast<AssignmentExpression>(
        ['var foo;', 'foo = 42;', 'foo = wrap(foo);'].join('\n'),
      );

      expect(
        resolveToValue(
          (path.get('right') as NodePath<CallExpression>).get('arguments')[0],
        ),
      ).toEqualASTNode(numericLiteral(42));
    });
  });

  describe('ImportDeclaration', () => {
    it('resolves unresolvable default import references to the import declaration', () => {
      const path = parse.expressionLast<Identifier>(
        ['import foo from "Foo"', 'foo;'].join('\n'),
      );
      const value = resolveToValue(path);

      expect(value.node.type).toBe('ImportDeclaration');
    });

    it('resolves unresolvable named import references to the import declaration', () => {
      const path = parse.expressionLast(
        ['import {foo} from "Foo"', 'foo;'].join('\n'),
      );
      const value = resolveToValue(path);

      expect(value.node.type).toBe('ImportDeclaration');
    });

    it('resolves unresolvable aliased import references to the import declaration', () => {
      const path = parse.expressionLast(
        ['import {foo as bar} from "Foo"', 'bar;'].join('\n'),
      );
      const value = resolveToValue(path);

      expect(value.node.type).toBe('ImportDeclaration');
    });

    it('resolves unresolvable namespace import references to the import declaration', () => {
      const path = parse.expressionLast(
        ['import * as bar from "Foo"', 'bar;'].join('\n'),
      );
      const value = resolveToValue(path);

      expect(value.node.type).toBe('ImportDeclaration');
    });

    it('resolves namespace import references to the import declaration', () => {
      const path = parse.expressionLast(
        `import * as bar from "Foo"; bar.baz`,
        mockImporter,
      );
      const value = resolveToValue(path);

      expect(value).toMatchSnapshot();
    });

    it('resolves namespace import references to the import declaration', () => {
      const path = parse.expressionLast(
        `import * as bar from "Foo"; bar['baz']`,
        mockImporter,
      );
      const value = resolveToValue(path);

      expect(value).toMatchSnapshot();
    });

    it('does not crash when resolving MemberExpression with non Identifiers', () => {
      const path = parse.expressionLast(
        `import * as bar from "Foo"; bar[()=>{}]`,
        mockImporter,
      );
      const value = resolveToValue(path);

      expect(value).toMatchSnapshot();
    });
  });

  describe('MemberExpression', () => {
    it("resolves a MemberExpression to it's init value", () => {
      const path = parse.expressionLast(
        ['var foo = { bar: 1 };', 'foo.bar;'].join('\n'),
      );

      expect(resolveToValue(path)).toEqualASTNode(numericLiteral(1));
    });

    it('resolves a MemberExpression in the scope chain', () => {
      const path = parse.expressionLast(
        ['var foo = 1;', 'var bar = { baz: foo };', 'bar.baz;'].join('\n'),
      );

      expect(resolveToValue(path)).toEqualASTNode(numericLiteral(1));
    });

    it('resolves a nested MemberExpression in the scope chain', () => {
      const path = parse.expressionLast(
        [
          'var foo = { bar: 1 };',
          'var bar = { baz: foo.bar };',
          'bar.baz;',
        ].join('\n'),
      );

      expect(resolveToValue(path)).toEqualASTNode(numericLiteral(1));
    });

    it('returns the last resolvable MemberExpression', () => {
      const path = parse.expressionLast(
        [
          'import foo from "bar";',
          'var bar = { baz: foo.bar };',
          'bar.baz;',
        ].join('\n'),
      );

      expect(resolveToValue(path)).toEqualASTNode(
        memberExpression(identifier('foo'), identifier('bar'), false, false),
      );
    });

    it('returns the path itself if it can not resolve it any further', () => {
      const path = parse.expressionLast(
        `var foo = {};
        foo.bar = 1;
        foo.bar;`,
      );

      expect(resolveToValue(path)).toEqualASTNode(path);
    });
  });
});
