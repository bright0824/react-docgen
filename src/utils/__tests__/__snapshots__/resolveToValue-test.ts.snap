// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveToValue ImportDeclaration does not crash when resolving MemberExpression with non Identifiers 1`] = `
Node {
  "computed": true,
  "object": Node {
    "name": "bar",
    "type": "Identifier",
  },
  "property": Node {
    "async": false,
    "body": Node {
      "body": Array [],
      "directives": Array [],
      "type": "BlockStatement",
    },
    "generator": false,
    "id": null,
    "params": Array [],
    "type": "ArrowFunctionExpression",
  },
  "type": "MemberExpression",
}
`;

exports[`resolveToValue ImportDeclaration resolves namespace import references to the import declaration 1`] = `
Node {
  "extra": Object {
    "raw": "3",
    "rawValue": 3,
  },
  "type": "NumericLiteral",
  "value": 3,
}
`;

exports[`resolveToValue ImportDeclaration resolves namespace import references to the import declaration 2`] = `
Node {
  "extra": Object {
    "raw": "3",
    "rawValue": 3,
  },
  "type": "NumericLiteral",
  "value": 3,
}
`;
